src/Modules/UngradedModules.jl

```diff
@@ -807,16 +908,211 @@ function base_ring(M::SubModuleOfFreeModule)
+  return M.F
+end
+
+@doc Markdown.doc"""
+    default_ordering(M::SubModuleOfFreeModule)
+
+Get the default ordering of `M`.
+"""
+function default_ordering(M::SubModuleOfFreeModule)
+  if !isdefined(M, :default_ordering)
+    try
+      ord = default_ordering(ambient_free_module(M))
+      set_default_ordering!(M, ord)
+      return M.default_ordering
+    catch err
+      error("Submodule has no default ordering")
+    end
+  end
+  return M.default_ordering
+end
+
+@doc Markdown.doc"""
+    set_default_ordering!(M::SubModuleOfFreeModule, ord::ModuleOrdering)
+
+Set the default ordering in `M` to `ord`.
+"""
+function set_default_ordering!(M::SubModuleOfFreeModule, ord::ModuleOrdering)
+  M.default_ordering = ord
+end
+
 @doc Markdown.doc"""
     std_basis(submod::SubModuleOfFreeModule)
 
-Compute a standard basis of `submod`.
+Compute a standard basis of `submod` with respect to its default odering.
 """
 function std_basis(submod::SubModuleOfFreeModule)
-  if !isdefined(submod, :std_basis)
-    submod.std_basis = groebner_basis(submod.gens)
+  return std_basis(submod, default_ordering(submod))
+end
+
+@doc Markdown.doc"""
+    groebner_basis(submod::SubModuleOfFreeModule)
+
+Compute a Gröbner basis of `submod` with respect to its default odering.
+For this, the default ordering must be global.
+"""
+function groebner_basis(submod::SubModuleOfFreeModule)
+  return groebner_basis(submod, default_ordering(submod))
+end
+
+function std_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+  if !haskey(submod.groebner_basis, ordering)
+    gb = compute_std_basis(submod, ordering)
+    submod.groebner_basis[ordering] = gb
   end
-  return submod.std_basis
+  return submod.groebner_basis[ordering]
+end
+
+@doc Markdown.doc"""
+    groebner_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+
+Compute a Gröbner of `submod` with respect to the given `ordering`.
+The ordering must be global.
+"""
+function groebner_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+  @assert is_global(ordering)
+  return std_basis(submod, ordering)
+end
+
+@doc Markdown.doc"""
+    reduced_groebner_basis(submod::SubModuleOfFreeModule)
+
+Compute a reduced Gröbner basis. The return type is `ModuleGens`.
+"""
+function reduced_groebner_basis(submod::SubModuleOfFreeModule)
+  return reduced_groebner_basis(submod, default_ordering(submod))
+end
+
+@doc Markdown.doc"""
+    reduced_groebner_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+
+Compute a reduced Gröbner basis with respect to the given `ordering`. The return type is `ModuleGens`.
+"""
+function reduced_groebner_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+  @assert is_global(ordering)
+  
+  if !haskey(submod.groebner_basis, ordering)
+    gb = compute_std_basis(submod, ordering, true)
+    submod.groebner_basis[ordering] = gb
+    return gb
+  else
+    if !(submod.groebner_basis[ordering].is_reduced)
```

Comment:
```suggestion
    submod.groebner_basis[ordering].is_reduced && return submod.groebner_basis[ordering]
```
Then again the rest of the code can be de-indented by one more level